File: src/main/java/NashornDemo.java
import jdk.nashorn.api.scripting.ScriptObjectMirror;
import org.apache.commons.io.IOUtils;

import javax.script.Invocable;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import java.time.LocalDateTime;
import java.util.Arrays;

public class NashornDemo {

    public static String fun1(String name) {
        System.out.format("Hi there from Java, %s", name);
        return "greetings from java";
    }

    public static void fun3(ScriptObjectMirror mirror) {
        System.out.println(mirror.getClassName() + ": " +
                Arrays.toString(mirror.getOwnKeys(true)));
    }

    public static void fun4(ScriptObjectMirror person) {
        System.out.println("Full Name is: " + person.callMember("getFullName"));
    }

    public static void main(String[] args) throws Exception {

        ScriptEngine engine = new ScriptEngineManager().getEngineByName("nashorn");
        final String script = IOUtils.toString(NashornDemo.class.getClassLoader().getResourceAsStream("script.js"), "UTF-8");
        engine.eval(script);
        Invocable invocable = (Invocable) engine;

        Object result = invocable.invokeFunction("fun1", "Peter Parker");
        System.out.println(result);
        System.out.println(result.getClass());

        invocable.invokeFunction("fun2", LocalDateTime.now());


    }
}

File: src/main/kotlin/NashornDemo.kt
import jdk.nashorn.api.scripting.ScriptObjectMirror
import org.apache.commons.io.IOUtils
import java.time.LocalDateTime
import java.util.*
import javax.script.Invocable
import javax.script.ScriptEngineManager

class NashornDemo {

    companion object {
        fun fun1(name: String): String {
            System.out.format("Hi there from Java, %s", name)
            return "greetings from java"
        }

        fun fun3(mirror: ScriptObjectMirror) {
            println(mirror.className + ": " +
                    Arrays.toString(mirror.getOwnKeys(true)))
        }

        fun fun4(person: ScriptObjectMirror) {
            println("Full Name is: " + person.callMember("getFullName"))
        }
    }
}

fun main(args: Array<String>) {

    val engine = ScriptEngineManager().getEngineByName("nashorn")
    val script = IOUtils.toString(NashornDemo::class.java.classLoader.getResourceAsStream("script.js"), "UTF-8")
    engine.eval(script)
    val invocable = engine as Invocable

    val result = invocable.invokeFunction("fun1", "Peter Parker")
    println(result)
    println(result.javaClass)

    invocable.invokeFunction("fun2", LocalDateTime.now())


}
